TNT-41044 Add telemetry reporting for server-side and hybrid requests

andyruwruw opened the pull request:
## Description

Currently we only send the TelemetryEntry to Delivery API for on-device-decisioning requests.  We should also report telemetry for server-side requests.

If the request includes a prefetch request then we can include telemetry inside the sendNotifications request.

## Related Issue

- [Jira Ticket](https://jira.corp.adobe.com/browse/TNT-41044)

dcottingham commented on packages/target-nodejs-sdk/src/index.js:
```diff
@@ -231,12 +241,23 @@ export default function bootstrap(fetchApi) {
         logger: this.logger,
         useBeacon: true,
-        ...options
+        ...request
```
Let's revert this back to options, since it's basically an envelope around the request + other metadata.  If we use the spread operator on request here then all the request fields will get assigned to targetOptions - which we don't want.

dcottingham commented on packages/target-nodejs-sdk/src/index.js:
```diff
@@ -231,12 +241,23 @@ export default function bootstrap(fetchApi) {
+      return executeDelivery(
+        targetOptions,
+        undefined,
+        this.telemetryProvider
```
Let's move telemetryProvider to be the 2nd param for executeDelivery so that in this case we can just send 2 args and not explicitly use `undefined` for decisioningEngine

dcottingham commented on packages/target-nodejs-sdk/src/index.js:
```diff
@@ -231,12 +241,23 @@ export default function bootstrap(fetchApi) {
+      ).then(preserveLocationHint.bind(this));
+    }
+
+    executeTelemetries(deliveryRequest, telemetryEntries) {
```
Let's move this fn to target-nodejs-sdk/helper.js to keep logic inside index.js as minimal as possible

andyruwruw replied:
that sounds a lot better

dcottingham commented on packages/target-tools/delivery-api-client/index.js:
```diff
@@ -1886,8 +1886,11 @@ class DeliveryApi extends BaseAPI {
+    async execute(imsOrgId, sessionId, deliveryRequest, version, telemetryProvider) {
+        const response = await this.executeRaw({ imsOrgId: imsOrgId, sessionId: sessionId, deliveryRequest: deliveryRequest, version: version, telemetryProvider: telemetryProvider });
+        
+        telemetryProvider.addEntry({});
```
Since this is generated code we shouldn't manually modify it.  Instead let's just wait until this fn returns to target.js and call addEntry() there

andyruwruw replied:
ooops good shout

andyruwruw commented on packages/target-decisioning-engine/test/decisioning.spec.js:
```diff
@@ -108,6 +126,10 @@ describe("decisioning engine", () => {
           expect(sendNotificationFunc.mock.calls.length).toEqual(1);
           const notificationPayload = sendNotificationFunc.mock.calls[0][0];
+
+          if (testDescription === "sends telemetry") {
```
Will fix, oops!

dcottingham commented on packages/target-nodejs-sdk/src/target.js:
```diff
@@ -122,7 +122,13 @@ export function executeDelivery(options, decisioningEngine) {
+      telemetryProvider
+    )
     .then((response = {}) => {
       logger.debug(
```
I think this would be an appropriate spot to call telemetryProvider.addEntry().  We have access to the response as well as the deliveryRequest and config.  So we should have all the data we need to create a telemetry entry.
For calculating the execution time we can use perftool.timeStart/timeEnd just like in decisionProvider.js

andyruwruw commented:
## Changes Made:

- TelemetryProvider moved to `target-nodejs-sdk` `index.js`
> To provide telemetries for both server-side and on-device decisioning, the TelemetryProvider needs to be available to both. server-side decisioning doesn't automatically handle sending notifications, so telemetry entries must persist between requests.
- DecisioningEngine no longer needs to know whether telemetries are enabled, the telemetry provider will handle that.
- NotificationProvider still maintains a TelemetryProvider, but it is instead passed in as a parameter.
- Guards to ensure telemetryProvider is defined in NotificationProvider before making calls on it.
> Possibly redundant from fixing tests. Might use this to remove some of the TelemetryProviders from tests.
- Added `mode` as property since it can be initialized when the TelemetryProvider is created. I think I might have jumped the gun on this.
- TelemetryProvider passed to every `executeDelivery` call. Made it second parameter since there are times when `decisioningEngine` isn't passed, but `telemetryProvider` is.
- TelemetryProvider passed to every `deliveryMethod` on `execute`.
- Each `deliveryMethod` creates a telemetry entry.
>Besides `execution` property, all requests should have entries.
- Added TelemetryProvider's to relevant tests, most with `noop` in place of `executeTelemetries`.
>The DecisioningEngine now expects a TelemetryProvider as a parameter, could alternatively handle `undefined` inside the DecisioningEngine, but figured this was more ideal given that every instance of the DecisioningEngine in the SDK separate from tests will have a TelemetryProvider. Could also just pass `undefined` in place of a TelemetryProvider when `telemetryEnabled` is `false`.

andyruwruw commented:
## Further Steps:

- Making coveralls happy.

dcottingham commented on packages/target-decisioning-engine/test/decisioning.spec.js:
```diff
@@ -89,15 +89,33 @@ describe("decisioning engine", () => {
 
-      const result = await decisioning.getOffers(input);
+      const result = await decisioning.getOffers({
+        ...input
```
Was there a particular reason you had to copy `input` into a new obj here?  Just curious

andyruwruw replied:
ah, side effect to how I used to have telemetryProvider passed in the same object. Decided that implementation sucked and switched it but forgot to fix this

dcottingham commented on packages/target-nodejs-sdk/src/index.js:
```diff
@@ -223,6 +235,10 @@ export default function bootstrap(fetchApi) {
 
       const visitor = createVisitor(options, this.config);
 
+      const request = this.telemetryProvider.executeTelemetries(
```
Nitpick:  I think for sake of clarity we should assign the resulting request back to options.request, just so it's more obvious that executeTelemetries is simply modifying the incoming options.request.  And then when assigning to targetOptions below we can just have `...options`

andyruwruw replied:
i'll disable the eslint error, it would be reassigning a parameter's property

dcottingham commented on packages/target-tools/delivery-api-client/index.js:
```diff
@@ -1888,6 +1888,7 @@ class DeliveryApi extends BaseAPI {
     }
     async execute(imsOrgId, sessionId, deliveryRequest, version) {
         const response = await this.executeRaw({ imsOrgId: imsOrgId, sessionId: sessionId, deliveryRequest: deliveryRequest, version: version });
+        
```
Let's remove this file from the PR since it's generated and there's only leftover whitespace

dcottingham commented on packages/target-tools/src/telemetryProvider.spec.js:
```diff
@@ -5,19 +5,28 @@ describe("TelemetryProvider", () => {
     execution: 1
   };
 
   it("adds an entry", () => {
```
Let's rename test something like "adds and executes entries" since it's more of an E2E test for the TelemetryProvider.  Let's also call `addEntry` 3x times before calling `executeTelemetries`.  Sometimes strange bugs present themselves when doing something more than once.

dcottingham commented on packages/target-nodejs-sdk/src/target.js:
```diff
@@ -45,6 +47,8 @@ export function executeDelivery(options, decisioningEngine) {
     useBeacon,
     createDeliveryApiMethod = createDeliveryApi
   } = options;
+  const timingTool = createPerfToolInstance();
```
Let's create timingTool outside this fn.  It should be a singleton, so we don't need to set it up for every executeDelivery call

andyruwruw replied:
I placed it at the top of the file!

dcottingham commented on packages/target-nodejs-sdk/src/target.js:
```diff
@@ -45,6 +47,8 @@ export function executeDelivery(options, decisioningEngine) {
     createDeliveryApiMethod = createDeliveryApi
   } = options;
+  const timingTool = createPerfToolInstance();
+  timingTool.timeStart(TIMING_EXECUTE_DELIVERY);
```
Let's move the timeStart call to the line before we call `deliveryMethod.execute` so we that we can measure just the time spent inside that function

dcottingham commented on packages/target-nodejs-sdk/src/target.js:
```diff
@@ -128,6 +132,13 @@ export function executeDelivery(options, decisioningEngine) {
+
+      if (deliveryMethod.decisioningMethod === DECISIONING_METHOD.SERVER_SIDE) {
+        telemetryProvider.addEntry(deliveryRequest, {
+          execution: timingTool.timeEnd(TIMING_EXECUTE_DELIVERY)
```
Let's move `timingTool.timeEnd(TIMING_EXECUTE_DELIVERY)` above the logger.debug statement so we measure as soon as we can.

dcottingham commented on packages/target-nodejs-sdk/src/target.js:
```diff
@@ -128,6 +132,13 @@ export function executeDelivery(options, decisioningEngine) {
         JSON.stringify(response, null, 2)
       );
+
+      if (deliveryMethod.decisioningMethod === DECISIONING_METHOD.SERVER_SIDE) {
```
Do you think we could remove the if statement and handle ODD telemetry here too?  Since inside the NotificationProvider the sendNotifications() call will call executeDelivery() for the ODD use case.  Feasible? Thoughts?

andyruwruw commented on packages/target-decisioning-engine/src/index.spec.js:
```diff
@@ -65,9 +65,12 @@ describe("TargetDecisioningEngine", () => {
   it("initializes", async () => {
     fetch.mockResponse(JSON.stringify(ARTIFACT_BLANK));
 
+    const telemetryProvider = TelemetryProvider(noop, null);
```
good advice, probably useless to reinstantiate it over and over, I'll do that for the other files as well

andyruwruw commented on packages/target-decisioning-engine/test/decisioning.spec.js:
```diff
@@ -89,15 +89,33 @@ describe("decisioning engine", () => {
 
       fetch.mockResponses(...mockResponses);
 
+      function executeTelementries(request, entries) {
```
Sure, I'll move it to `target-tools` `utils` and export it through so both of the other packages can access it.

dcottingham commented on packages/target-nodejs-sdk/test/telemetry.spec.js:
```diff
@@ -0,0 +1,422 @@
+      });
+    });
+
+    it("executes telemetries when server-side used", () => {
```
Let's rename this to `executes telemetries on next getOffers call when server-side used` so it's a bit more clear

dcottingham commented on packages/target-nodejs-sdk/test/telemetry.spec.js:
```diff
@@ -0,0 +1,422 @@
+  });
+
+  describe("hybrid", () => {
+    it("calls sendNotifications when on-device used", () => {
```
Let's rename this to `executes telemetries as part of sendNotifications call when on-device used` so it's a bit more clear

dcottingham commented on packages/target-nodejs-sdk/test/telemetry.spec.js:
```diff
@@ -0,0 +1,422 @@
+import { DECISIONING_METHOD, executeTelemetries } from "@adobe/target-tools";
```
Great job on the test cases!  Covers all the use cases.

andyruwruw commented on packages/target-nodejs-sdk/test/telemetry.spec.js:
```diff
@@ -0,0 +1,422 @@
+jest.mock("@adobe/target-tools", () => ({
+  ...jest.requireActual("@adobe/target-tools"),
+  executeTelemetries: jest.fn(
+    jest.requireActual("@adobe/target-tools").executeTelemetries
```
It turns it into a mock function, but keeps the same implementation. So although it's now a mock function, it still adds `telemetry.entries` to a request. I believe there is a check later down the road that will throw an error if this function doesn't perform its task. `jest.requireActual` just asks jest to go and find the original `executeTelemetries` even though we're redefining the function.

dcottingham replied:
cool, that's good to know

andyruwruw commented on packages/target-tools/src/telemetryProvider.js:
```diff
@@ -8,23 +8,23 @@ import { DECISIONING_METHOD } from "./enums";
    * @param {import("@adobe/target-tools/delivery-api-client/models/TelemetryEntry").TelemetryEntry} entry
    */
-  function addEntry(entry) {
+  function addEntry(request, entry = {}, decisioningMethod = mode) {
```
I think before I was just dumb and didn't realize the `execution` metric applied to all requests, so I was passing in empty entries while writing. Very true though, I can remove that now that we always send data
